docs/source/_repopath.py

```diff
@@ -0,0 +1,84 @@
+import functools
+import logging
+import os
+import subprocess
+from pathlib import Path
+from docutils import nodes
+
+from sphinx.application import Sphinx as SphinxApp
+from sphinx.roles import EmphasizedLiteral
+
+
+_logger = logging.getLogger(f"sphinx.ext.{__name__}")
+
+
+@functools.lru_cache
+def _get_repo_root() -> Path:
+    try:
+        repo_root = subprocess.check_output(
+            ["git", "rev-parse", "--show-toplevel"],
+            text=True,
+        ).strip()
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError("Unable to get repository root") from e
+    return Path(repo_root).resolve()
+
+
+class RepoPath(EmphasizedLiteral):
+    @property
+    def path(self) -> Path:
+        repo_root = _get_repo_root()
+        return repo_root / self.text
+
+    def run(self):
+        if not self.path.exists():
+            _logger.warning("path %r does not exist", self.text)
+        return super().run()
+
+
+class PathRoleBase:
+    def __init__(self, base_path=""):
+        self.base_path = Path(base_path)
```

Comment:
why the Path constructor here? all the usage already passes a Path into the PathRoleBase

Reply:
Mostly for [robustness](https://en.wikipedia.org/wiki/Robustness_principle)/defensiveness: the internals only work if `self.base_path` is a `pathlib.Path`, and `pathlib.Path` instances are immutable so creating a new `Path` is easier than having `isinstance(..., Path)` checks.